### PR TITLE
Truncate endTime to seconds in TimeSeries range updates.

### LINKED ISSUE #94470: [CI] TSDBIndexingIT testTimeRanges failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/r727kot4zv3qm/tests/:modules:data-streams:internalClusterTest/org.elasticsearch.datastreams.TSDBIndexingIT/testTimeRanges

**Reproduction line:**
```
./gradlew ':modules:data-streams:internalClusterTest' --tests "org.elasticsearch.datastreams.TSDBIndexingIT.testTimeRanges" -Dtests.seed=2DDC294DD605AC70 -Dtests.locale=de -Dtests.timezone=America/Argentina/La_Rioja -Druntime.java=19
```

**Applicable branches:**
main

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.datastreams.TSDBIndexingIT&tests.test=testTimeRanges

**Failure excerpt:**
```
org.elasticsearch.index.mapper.MapperParsingException: failed to parse

  at __randomizedtesting.SeedInfo.seed([2DDC294DD605AC70:EB774AA41A1D47C2]:0)
  at org.elasticsearch.index.mapper.DocumentParser.wrapInMapperParsingException(DocumentParser.java:212)
  at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:70)
  at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:78)
  at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:1021)
  at org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:969)
  at org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:913)
  at org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:354)
  at org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:219)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:286)
  at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:137)
  at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:74)
  at org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:214)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:958)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  at java.lang.Thread.run(Thread.java:1589)

  Caused by: java.lang.IllegalArgumentException: time series index @timestamp value [2023-03-10T17:05:54.863Z] must be smaller than 2023-03-10T17:05:54.359Z

    at org.elasticsearch.index.mapper.DataStreamTimestampFieldMapper.validateTimestamp(DataStreamTimestampFieldMapper.java:258)
    at org.elasticsearch.index.mapper.DataStreamTimestampFieldMapper.postParse(DataStreamTimestampFieldMapper.java:223)
    at org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:114)
    at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:67)
    at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:78)
    at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:1021)
    at org.elasticsearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:969)
    at org.elasticsearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:913)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:354)
    at org.elasticsearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:219)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnPrimary(TransportShardBulkAction.java:286)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:137)
    at org.elasticsearch.action.bulk.TransportShardBulkAction.dispatchedShardOperationOnPrimary(TransportShardBulkAction.java:74)
    at org.elasticsearch.action.support.replication.TransportWriteAction$1.doRun(TransportWriteAction.java:214)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:958)
    at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
    at java.lang.Thread.run(Thread.java:1589)

```

### PR DESCRIPTION
**Pull request by kkrik-es**
This change is in line with the DataStream logic for picking the right 
backing index to write a doc based on its timestamp. Without truncating, 
inserted docs can be directed to an index with endTime earlier than their 
timestamp, triggering validation errors.

Resolves #94470

### CODE REVIEW

**Review by martijnvg — APPROVED** [review]
I left one comment, otherwise LGTM.
The test failure looks unrelated.

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java [hunk 1]
```diff
@@ -1054,4 +1054,8 @@ public int hashCode() {
             return Objects.hash(name);
         }
     }
+
+    public static Instant getCanonicalTimestampBound(Instant i) {
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:1058]: and maybe rename `i` to `time` or `untruncated`?

[the remark this answers is not in the document]
Comment (kkrik-es) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java]: Yes! Added, please take another look.

**Review by martijnvg — APPROVED** [review]
LGTM!

**Comment by kkrik-es** [comment]
Thanks for the quick review! qq I don't see a yaml file, am I supposed to generate it manually and add it here?

**Comment by martijnvg** [comment]
> Thanks for the quick review! qq I don't see a yaml file, am I supposed to generate it manually and add it here?

Good point. Normally the bot automatically adds a changelog entry. I think in this case the changelog entry should be added to the backport.

**Pull request merged**